lib/posix/pthread_key.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <misc/dlist.h>
+#include <posix/pthread.h>
+
+struct k_sem pthread_key_sem;
+
+typedef struct pthread_thread_data {
+	sys_dnode_t node;
+
+	/* Key and thread specific data passed to pthread_setspecific() */
+	pthread_key_obj *key;
+	void *spec_data;
+
+} pthread_thread_data;
+
+typedef struct pthread_key_data {
+	sys_dnode_t node;
+	pthread_thread_data thread_data;
+} pthread_key_data;
+
+
+/**
+ * @brief Create a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_create(pthread_key_t *key,
+		void (*destructor)(void *))
+{
+	pthread_key_obj *new_key;
+
+	*key = NULL;
+
+	k_sem_init(&pthread_key_sem, 1, 1);
```

Comment:
Why is pthread_key_sem() initialized every time pthread_key_create() is called?

Reply:
Okay, so my understanding is that, for every new key we create, we will need to initialize a semaphore, which will be taken and given when we modify this key's (the respective key created with `pthread_key_create`) thread list using the pthread_get_specific, etc functions.

And this is where `pthread_once()` is useful - we will pass `pthread_key_create()` to `pthread_once()` so that the function is called only once for a particular key creation. So for the same key, there won't be multiple initialization of the semaphore.

Reply:
There's just one semaphore for all keys, and it's being initialized every time. It's not going to work.

Reply:
Okay, I got your point, I've statically initialized now